Add support for nested Aws StepFunctions service integration

ca-nguyen opened the pull request:
*Issue #, if available:* #125

*Description of changes:*
Added support for [StartExecution](https://docs.aws.amazon.com/step-functions/latest/dg/connect-stepfunctions.html) step that allows to start an execution of another state machine

*Testing*
- Added unit tests
- Manual tests for all resources:
   - "arn:aws:states:::states:startExecution"
   - "arn:aws:states:::states:startExecution.sync:2"
   - "arn:aws:states:::states:startExecution.sync"
   - "arn:aws:states:::states:startExecution.waitForTaskToken"


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,59 @@ def __init__(self, state_id, **kwargs):
+                kwargs[Field.Resource.value] = get_service_integration_arn(STEP_FUNCTIONS_SERVICE_NAME,
+                                                                       StepFunctions.StartExecution,
+                                                                       IntegrationPattern.WaitForCompletion)
+        else:
+            """
+            Example resource arn:aws:states:::states:startExecution
+            """
+            kwargs[Field.Resource.value] = get_service_integration_arn(STEP_FUNCTIONS_SERVICE_NAME,
+                                                                       StepFunctions.StartExecution)
```
suggestion: should there be a flag to ask for this? 

Currently, Asking for this property is also counter-intuitive because it's a behaviour that is achieved by setting `wait_for_callback` and `wait_for_completion` to false. 

A customer saying "no" to these 2 properties != a "yes" for the behaviour they will get. Additionally, this behaviour is not currently documented in any way.

Recommendation:
* Add a flag to opt into the behaviour. Clearly document what happens when that flag is set to `True`
  *   Alternatively could model these 3 properties in a more complex type, but I'm not leaning that way because that might be disruptive and too different from usage of other service integrations.

ca-nguyen replied:
I like your suggestion - I agree, having to disable to properties to activate another is not intuitive. A flag might make things clearer!
I had thought of introducing an enum for the property type, but as you mentioned, the implementation would diverge from the other service integrations.
I can add a check to see if more than 1 is enabled and throw an exception if that is the case.

ca-nguyen replied:
What would be the default if all flags are disabled? Should we return an error?
If we were to keep only the 2 initial property flags `wait_for_callback` and `wait_for_completion` and:
- Raise an exception if both were set to True (as you mentioned below) 
- Update documentation to illustrate how to select this behaviour (Disable both false to get arn:aws:states:::states:startExecution resource)

I think this way, would be clearer for the customers instead of having them enable at least 1 of the 3 flags - What do you think?

shivlaks replied:
> If we were to keep only the 2 initial property flags wait_for_callback and wait_for_completion

How would users use the async invocation method, this only enables 2/3 possible modes of invoking the service integration. I think we need to provide users with an intuitive way to ask for `arn:aws:states:::states:startExecution`.

I'm not in favour of setting `wait_for_callback` and `wait_for_completion` to false to achieve that. It's counter-intuitive to me as a developers, but perhaps others will disagree...


* the default if none of the properties are set should be to use `wait_for_completion` aka `sync:2`
  * Rationale: it's what all the other service integrations establish as a default
* if a user sets all the flags to False, then it's a validation error
  * Rationale: it's required that one of the modes of operation is established and the user's intention is unclear.

Aside:
These implementation decisions should be captured and elaborated in the PR summary itself. Right now the description only captures the outcome, but it's valuable context for posterity as well as helping reviewers understand why the code is structured the way it is. Please ensure you update the PR description with implementation decisions that are being made.

ca-nguyen replied:
>- if a user sets all the flags to False, then it's a validation error
Rationale: it's required that one of the modes of operation is established and the user's intention is unclear.

Agreed - this will ensure the customer's intention is clear. 
I'll go with what you are proposing - we want to provide an intuitive solution for our customers

ca-nguyen replied:
>Please ensure you update the PR description with implementation decisions that are being made.

Updated the description

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,59 @@ def __init__(self, state_id, **kwargs):
+        """
+        Args:
+            state_id (str): State name whose length **must be** less than or equal to 128 unicode characters. State names **must be** unique within the scope of the whole state machine.
+            wait_for_callback(bool, optional): Boolean value set to `True` if the Task state should wait for callback to resume the operation. (default: False)
```
nit: we should probably keep all the service integration pattern flags co-located in the docs. locality is valuable and easier to digest. I think we can do better by not having related properties (like wait_for_completion) documented 10 lines apart.

ca-nguyen replied:
I agree - will regroup them in the next commit

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,59 @@ def __init__(self, state_id, **kwargs):
+            wait_for_completion (bool, optional): Boolean value set to `True` if the Task state should wait to complete before proceeding to the next step in the workflow. (default: True)
+            json_output(bool, optional): Boolean value set to `True` if the the Task should return a response in JSON format. If set to `False`, the Task should return a str response (default: False)
+        """
+        if wait_for_callback:
+            """
+            Example resource arn: arn:aws:states:::states:startExecution.waitForTaskToken
+            """
+
+            kwargs[Field.Resource.value] = get_service_integration_arn(STEP_FUNCTIONS_SERVICE_NAME,
+                                                                       StepFunctions.StartExecution,
+                                                                       IntegrationPattern.WaitForTaskToken)
+        elif wait_for_completion:
```
missing validation:

if both `wait_for_callback` and `wait_for_completion` properties are set `True`, what the customer wants is unclear and we should raise an error. What will actually happen is we will *always* give them `wait_for_callback`

ca-nguyen replied:
Agreed - will add that in the next commit

shivlaks reviewed (changes requested):
left some feedback of some changes I think we should make and some suggestions to consider. haven't looked at the tests yet.

ca-nguyen commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,59 @@ def __init__(self, state_id, **kwargs):
+            result_path (str, optional): Path specifying the raw input’s combination with or replacement by the state’s result. (default: '$')
+            output_path (str, optional): Path applied to the state’s output after the application of `result_path`, producing the effective output which serves as the raw input for the next state. (default: '$')
+            wait_for_completion (bool, optional): Boolean value set to `True` if the Task state should wait to complete before proceeding to the next step in the workflow. (default: True)
+            json_output(bool, optional): Boolean value set to `True` if the the Task should return a response in JSON format. If set to `False`, the Task should return a str response (default: False)
```
>- default to sync:2

Agreed - let's default to sync:2 - I think we should be consistent with the CDK and this seems like it would be the most used resource between the two.

>- don't make json_output configurable

Since `startExecution.sync` is in our documentation, I believe it could be good to provide an option for our customers to use it. How about instead of being `json_output` flag, we provide a flag to opt in for a string response instead? Something like `string_response` or `string_output` that can be set in the kwargs. I'm curious to see what other reviewers think too and will go with the majority

shivlaks replied:
> Since startExecution.sync is in our documentation, I believe it could be good to provide an option for our customers to use it.

It will always be in the documentation as it was the only option available initially. As we refined our approach and offered a better alternative, `sync:2` was introduced. 

We shouldn't seek to hit arbitrarily high 1:1 parity with documentation. Rather, we should ask ourselves what use cases it unlocks, whether `sync:2` precludes the ability to do anything that `sync` offers. Do you have a use case in mind?

We have not exposed this integration to customers yet, so my recommendation is still to start as tight as possible and expand it in the future if we have a feature request / use case in mind. Nothing prevents us from exposing it in a future release, but supporting it at the outset based on assumptions is fragile and also a 1-way door as we'll have it out there forever (at least until v3).

thoughts?

ca-nguyen replied:
>We have not exposed this integration to customers yet, so my recommendation is still to start as tight as possible and expand it in the future if we have a feature request / use case in mind. Nothing prevents us from exposing it in a future release, but supporting it at the outset based on assumptions is fragile and also a 1-way door as we'll have it out there forever (at least until v3).

I definitely agree with this - we shouldn't introduce 1 way doors. I don't have as specific use case in mind where `:sync` would be preferred over `:sync:2`. If there is ask for it, we can always add support for it later and it won't be a breaking change

shivlaks replied:
> we shouldn't introduce 1 way doors.

That's not quite what I meant. I meant that when we go through a 1-way door, we should have a deep understanding of **why** we are doing it.

wong-a commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,74 @@ def __init__(self, state_id, **kwargs):
+        # Validate that only 1 one of the property flags are set to true
+        enabled_property_flags = []
+        property_flags = [(wait_for_callback, "wait_for_callback"), (wait_for_completion, "wait_for_completion")]
+        for property_flag in property_flags:
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)

wong-a commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,66 @@ def __init__(self, state_id, **kwargs):
+        property_flags = [(wait_for_callback, "wait_for_callback"), (wait_for_completion, "wait_for_completion"),
+                          (async_call, "async_call")]
+        enabled_property_flags = [property_flag[1] for property_flag in property_flags if property_flag[0]]
+        if len(enabled_property_flags) == 0:
```
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0` or `if len(val) &gt; 0`

 [Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

wong-a commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +892,66 @@ def __init__(self, state_id, **kwargs):
+            state_id (str): State name whose length **must be** less than or equal to 128 unicode characters. State names **must be** unique within the scope of the whole state machine.
+            wait_for_callback(bool, optional): Boolean value set to `True` if the Task state should wait for callback to resume the operation. (default: False)
+            wait_for_completion (bool, optional): Boolean value set to `True` if the Task state should wait to complete before proceeding to the next step in the workflow. (default: True)
+            async_call(bool, optional): Boolean value set to `True` for the Task proceed to the next step in the workflow without waiting for completion. (default: False):
```
This would be a good opportunity to improve as proposed here: https://github.com/aws/aws-step-functions-data-science-sdk-python/pull/74

Using a single enum arg is easier to reason about instead of three mutually exclusive args, which either overwrite each other or raise an Error if more than one are provided.

shivlaks replied:
I'm in favour of an enum as well (which is what we also do in the CDK) - the tradeoff is it might be dissimilar to other service integrations.

i.e. if any of the existing service integrations add support for all 3 modes of operation, we would likely have some inconsistency between service integrations.

not sure that it's a problem we need to futureproof for, but just mentioning it for completeness

ca-nguyen replied:
Also in favour of an enum even though it introduces inconsistency with the other service integrations
This will be clearer and easier to use for our customers

wong-a replied:
None of the service integrations today support all 3 patterns, which is why this is a good time to introduce this change.

`ecs:runTask` does support all 3, but this SDK only exposes `.sync` and request-response. 

- https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/main/src/stepfunctions/steps/compute.py#L184-L197
- https://docs.aws.amazon.com/step-functions/latest/dg/connect-ecs.html

shivlaks replied:
this is the enum we added in the [CDK](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-stepfunctions.IntegrationPattern.html)
and Python specific [documentation](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_stepfunctions/IntegrationPattern.html) for it

keep in mind that you'd want to name the options in the enum in a way that's idiomatic for existing data science SDK customers.

wong-a replied:
We have an enum already https://github.com/aws/aws-step-functions-data-science-sdk-python/blob/23878de4ba909564569b90d9ab3ae689cacf04dd/src/stepfunctions/steps/integration_resources.py#L20

ca-nguyen replied:
I was planning to use the the existing enum mentioned by @wong-a 
Change to use enum arg instead of 3 flags will be included in the next commit

shivlaks replied:
i should clarify:

wasn't suggesting adding a new enum, was only sharing the CDK enum to indicate that there shouldn't be a new entry in the enum to surface `sync:2` - it's just what `WaitForCompletion` maps to for this service integration.

> None of the service integrations today support all 3 patterns, which is why this is a good time to introduce this change.

That's true, but my comment around futureproofing is more around how you'd introduce it if a service integration that currently supports 2 patterns starts supporting all 3.

i.e. if EMR aded wait for completion, since we have already exposed `wait_for_completion` as a property:
* we could introduce the enum as input allowing for all 3 integration types
  * we would essentially map 2 properties as ways in which wait_for_completion could be requested (enum and flag)
  * we would need to add validation that enum and this flag don't conflict with each other

I still prefer the enum because it's clear and explicit about behaviour that will be observed.

ca-nguyen replied:
>wasn't suggesting adding a new enum, was only sharing the CDK enum to indicate that there shouldn't be a new entry in the enum to surface sync:2 - it's just what WaitForCompletion maps to for this service integration.

Yes, i was planning to remove the `WaitForCompletionWithJsonResponse` that was added in this PR to not expose the response format to the customer (this was discussed [here](https://github.com/aws/aws-step-functions-data-science-sdk-python/pull/166#discussion_r706701234))

ca-nguyen replied:
After reflexion, I opted to add a new enum so that we have the same terms than the ones described in [Service Integration Patterns](https://docs.aws.amazon.com/step-functions/latest/dg/connect-to-resource.html) similarly to the CDK doc mentioned by @shivlaks (details in PR description)

ca-nguyen replied:
>i.e. if EMR aded wait for completion, since we have already exposed wait_for_task_token as a property:
> - we could introduce the enum as input allowing for all 3 integration types
>   - we would essentially map 2 properties as ways in which wait_for_task_token could be requested (enum and flag)
>   - we would need to add validation that enum and this flag don't conflict with each other
> I still prefer the enum because it's clear and explicit about behaviour that will be observed.
> If that situation materialized, I'd also recommend marking the flag as deprecated and dropping support in v3 to normalize every integration to using the enum.

This seems like an acceptable solution when adding support for a 3rd integration pattern
Mapping the existing flag to the enum and marking that flag as "to be deprecated" will ensure backward compatibility and give a heads up to our customers

For consistency, when adding support for a new service integrations, regardless of the number of integration patterns, I also think we should use the enum solution if we decide to go that route

shivlaks commented:
not directly related to this PR, but since we are trying to address the suggestion in #74 - any reason we shouldn't follow the same pattern for un-released service integrations we've added recently in #156 and #151 (not in this PR)

I feel we should establish that in all service integrations we support, but especially ones we have not released yet.

thoughts? - @wong-a  @ca-nguyen

ca-nguyen commented:
> not directly related to this PR, but since we are trying to address the suggestion in #74 - any reason we shouldn't follow the same pattern for un-released service integrations we've added recently in #156 and #151 (not in this PR)

Since these service integrations have not been released yet, it makes sense to follow the same patterns that provide a clearer solution to our customers. 

>I feel we should establish that in all service integrations we support, but especially ones we have not released yet.
I agree - ideally, we are consistent through all the service integrations

shivlaks commented on src/stepfunctions/steps/integration_resources.py:
```diff
@@ -25,6 +29,17 @@ class IntegrationPattern(Enum):
+    """
+
+    REQUEST_RESPONSE = "RequestResponse",
+    RUN_A_JOB = "RunAJob",
```
this is just called `RUN_JOB` in the CDK. I would think that we should try and stick to something similar?

shivlaks reviewed (changes requested):
not sure I follow this part tradeoffs:
>  An alternative mentioned by @shivlaks (in the discussion) is to use the existing flags to map to the ServiceIntegrationType and add a note in the docstrings that those flags will be deprecated in v3.0.0 and deprecate them when we are ready to upgrade the major version

I was not suggesting using existing flags to map to service integration type. I was indicating that we should *not* have default behaviour that a user does not explicitly opt into. The callout was that *if* a user achieves a certain behaviour, they need to explicitly opt into it.

I don't think this is a design decision at all and really the only tradeoff worth documenting is that we are opting into inconsistency for simplicity. The future consideration worth documenting is what we will do with existing service integrations as well as all future integrations. Will other service integrations now follow the enum pattern or fall back to past patterns?

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +893,60 @@ def __init__(self, state_id, **kwargs):
+            result_path (str, optional): Path specifying the raw input’s combination with or replacement by the state’s result. (default: '$')
+            output_path (str, optional): Path applied to the state’s output after the application of `result_path`, producing the effective output which serves as the raw input for the next state. (default: '$')
+        """
+        supported_integ_types = [ServiceIntegrationType.REQUEST_RESPONSE, ServiceIntegrationType.RUN_A_JOB,
```
you called the enum patterns

```suggestion
        supported_integration_patterns = [ServiceIntegrationType.REQUEST_RESPONSE, ServiceIntegrationType.RUN_A_JOB,
```

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +893,60 @@ def __init__(self, state_id, **kwargs):
+        supported_integ_types = [ServiceIntegrationType.REQUEST_RESPONSE, ServiceIntegrationType.RUN_A_JOB,
+                                 ServiceIntegrationType.WAIT_FOR_CALLBACK]
+
+        if not isinstance(service_integration_type, ServiceIntegrationType):
+            raise ValueError(f"Invalid type used for service_integration_type arg ({service_integration_type}, "
+                             f"{type(service_integration_type)}). Accepted type: {ServiceIntegrationType}")
+        elif service_integration_type not in supported_integ_types:
+            raise ValueError(f"Service Integration Type ({service_integration_type.name}) is not supported for this step - "
+                             f"Please use one of the following: "
+                             f"{[integ_type.name for integ_type in supported_integ_types]}")
```
can we move the validation into it's own function and keep the constructor clean??
I imagine you'll want to reuse this validation across other service integrations as well.

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +893,60 @@ def __init__(self, state_id, **kwargs):
+                             f"{[integ_type.name for integ_type in supported_integ_types]}")
+
+        if service_integration_type == ServiceIntegrationType.RUN_A_JOB:
+            # For RunAJob service integration type, use the resource that returns json response (`.sync:2`)
```
nit: unnecessary comment - if we do want to document, we should capture the "why". the code already explains what it's doing.

shivlaks commented on tests/unit/test_service_steps.py:
```diff
@@ -1158,3 +1161,92 @@ def test_eks_call_step_creation():
+
+
+@patch.object(boto3.session.Session, 'region_name', 'us-east-1')
+def test_step_functions_start_execution_step_creation():
```
test name should include the integration pattern you're testing or "default" if you make it optional (as it is in the CDK)

shivlaks commented on src/stepfunctions/steps/service.py:
```diff
@@ -887,3 +893,60 @@ def __init__(self, state_id, **kwargs):
+        """
+        Args:
+            state_id (str): State name whose length **must be** less than or equal to 128 unicode characters. State names **must be** unique within the scope of the whole state machine.
+            integration_pattern (stepfunctions.steps.integration_resources.ServiceIntegrationType): Service integration type to use to build resource.
```
1. should this be optional with a default?
2. this is not the name of your parameter

shivlaks commented on tests/unit/test_service_steps.py:
```diff
@@ -1158,3 +1161,92 @@ def test_eks_call_step_creation():
+    step = StepFunctionsStartExecutionStep(
+        "SFN Start Execution", service_integration_type=ServiceIntegrationType.REQUEST_RESPONSE, parameters={
+            "Input": {
+                "Comment": "Hello world!"
```
nit: is comment needed at all? - try to keep tests minimal. applies to all tests

shivlaks reviewed (changes requested):
getting closer and certainly looks a lot nicer with the integration pattern. 
take a look at my latest round of feedback and let me know if you have any questions

ca-nguyen commented:
>But no integ tests are in the diff. Did you forget to commit them? What kind of manual testing did you do?

I added details on the Manual tests that were performed in the `Testing` section of the PR summary.
No integ test were added since we needed an external Lambda function to test. This was checked since this was N/A and to reflect that no more tasks were left. The template says:

>Please check all boxes (including N/A items)

I updated with a Note next to the checked task to avoid confusion
